### ISSUE TITLE
Metric Parameter not Working as Intended

### ISSUE THREAD

**Issue by jcoding2022** [microsoft/LightGBM#6011]
## Description

In modeling a binary classification problem, the results do not change with the "metric" parameter.

## Reproducible example

```
import lightgbm as lgb
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

# Load the breast cancer dataset (example dataset)
data = load_breast_cancer()
X = data.data
y = data.target

# Split the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define the LightGBM classifier
lgb_model = lgb.LGBMClassifier(metric='binary_logloss')  # Using binary log-loss as the evaluation metric

# Fit the model to the data
lgb_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = lgb_model.predict(X_test)

# Evaluate the model using the specified metric
accuracy = lgb_model.score(X_test, y_test)
print(f"Accuracy: {accuracy:.4f}")
```

I tried to change the metric parameter from 'binary_logloss" to 'auc', 'f1', or 'average_precision', the results kept the same.

## Environment info

LightGBM version or commit hash: 4.0.0

Command(s) you used to install LightGBM

```
pip install lightgbm

```

## Additional Comments

Did I miss anything here?

**Comment by jmoralez** [comment]
Hey @jcoding2022, thanks for using LightGBM.

I think you mean to use [objective](https://lightgbm.readthedocs.io/en/latest/Parameters.html#objective). The metric is used when you're computing evaluation metrics, e.g. when using `valid_sets` and the [log_evaluation callback](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.log_evaluation.html#lightgbm.log_evaluation).

Please let us know if you have further doubts.

**Comment by jcoding2022** [comment]
> Hey @jcoding2022, thanks for using LightGBM.
> 
> I think you mean to use [objective](https://lightgbm.readthedocs.io/en/latest/Parameters.html#objective). The metric is used when you're computing evaluation metrics, e.g. when using `valid_sets` and the [log_evaluation callback](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.log_evaluation.html#lightgbm.log_evaluation).
> 
> Please let us know if you have further doubts.

Thanks for your quick reply. Can you point to an example that shows how the metric parameter works?

**Comment by jcoding2022** [comment]
> Hey @jcoding2022, thanks for using LightGBM.
> 
> I think you mean to use [objective](https://lightgbm.readthedocs.io/en/latest/Parameters.html#objective). The metric is used when you're computing evaluation metrics, e.g. when using `valid_sets` and the [log_evaluation callback](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.log_evaluation.html#lightgbm.log_evaluation).
> 
> Please let us know if you have further doubts.

If my objective is to increase F1_score, what parameters and values should I use?

**Comment by jameslamb** [comment]
> Can you point to an example that shows how the metric parameter works?

https://github.com/microsoft/LightGBM/blob/e9fface2f7013014a4100270b01f44efd2b920d3/examples/python-guide/sklearn_example.py#L24-L30

> If my objective is to increase F1_score, what parameters and values should I use?

If you want to continue using the `scikit-learn` interface to LightGBM:

* use `LGBMClassifier()`
   - *it will automatically set `objective="binary"` when passed a label with 2 unique values*
* define a custom eval metric as a Python function, and pass it through the `eval_metric` parameter to `LGBMClassifier.fit()`

Consider the following minimal example, which I tested with `lightgbm` v4.0.0

```python
import lightgbm as lgb
from sklearn.datasets import load_breast_cancer
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split

# Load a binary classification dataset
X, y = load_breast_cancer(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


# define F1 score as a custom metric
def f1_metric(y_true, y_pred):
    metric_name = "F1"
    # NOTE: y_pred will be a probability of the majority class
    value = f1_score(y_true, y_pred > 0.5)
    higher_better = True
    return metric_name, value, higher_better


# Fit the model to the data
# NOTE: max_depth, num_leaves, min_child_samples set to
#       smaller-than-default values because the dataset
#       only has around 450 rows
lgb_model = lgb.LGBMClassifier(
    verbosity=1,
    n_estimators=10,
    max_depth=3,
    num_leaves=7,
    min_child_samples=2
)

lgb_model.fit(
    X=X_train,
    y=y_train,
    eval_set=[(X_test, y_test)],
    eval_names="valid",
    eval_metric=["auc", "binary_error", f1_metric],
    callbacks=[lgb.log_evaluation(1)]
)
```

You should see all of those metrics passed to `eval_metric` printed at each iteration.

```text
[LightGBM] [Info] Number of positive: 286, number of negative: 169
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000043 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 4548
[LightGBM] [Info] Number of data points in the train set: 455, number of used features: 30
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.628571 -> initscore=0.526093
[LightGBM] [Info] Start training from score 0.526093
[1]	valid's auc: 0.956436	valid's binary_error: 0.377193	valid's binary_logloss: 0.582173	valid's F1: 0.767568
[2]	valid's auc: 0.972977	valid's binary_error: 0.377193	valid's binary_logloss: 0.514326	valid's F1: 0.767568
[3]	valid's auc: 0.971667	valid's binary_error: 0.0526316	valid's binary_logloss: 0.462733	valid's F1: 0.959459
[4]	valid's auc: 0.97265	valid's binary_error: 0.0526316	valid's binary_logloss: 0.415805	valid's F1: 0.958904
[5]	valid's auc: 0.97265	valid's binary_error: 0.0438596	valid's binary_logloss: 0.37737	valid's F1: 0.965517
[6]	valid's auc: 0.967082	valid's binary_error: 0.0438596	valid's binary_logloss: 0.347271	valid's F1: 0.965517
[7]	valid's auc: 0.966754	valid's binary_error: 0.0350877	valid's binary_logloss: 0.319601	valid's F1: 0.972222
[8]	valid's auc: 0.967245	valid's binary_error: 0.0350877	valid's binary_logloss: 0.294086	valid's F1: 0.972222
[9]	valid's auc: 0.994432	valid's binary_error: 0.0350877	valid's binary_logloss: 0.268995	valid's F1: 0.972222
[10]	valid's auc: 0.994104	valid's binary_error: 0.0350877	valid's binary_logloss: 0.251823	valid's F1: 0.972222
```

Note that the F1 score on the held-out validation set is improving with each iteration.

Try modifying `eval_name`, `eval_metrics`, and the number of iterations passed through `lgb.log_evaluation()` to customize this.

You can recover all of the metric values in the `evals_result_` property.

```python
lgb_model.evals_result_
```

```text
{'valid': OrderedDict([('auc', [0.9564362921716345, 0.9729773992793973, 0.9716672125777923, 0.9726498526039961, 0.9726498526039961, 0.9670815591221749, 0.9667540124467736, 0.9672453324598755, 0.9944317065181788, 0.9941041598427776]), ('binary_error', [0.37719298245614036, 0.37719298245614036, 0.05263157894736842, 0.05263157894736842, 0.043859649122807015, 0.043859649122807015, 0.03508771929824561, 0.03508771929824561, 0.03508771929824561, 0.03508771929824561]), ('binary_logloss', [0.5821729220177619, 0.5143256933774838, 0.4627328015843806, 0.415804786329259, 0.3773696159284982, 0.3472708147431134, 0.31960105920456877, 0.29408603472697215, 0.26899508360230007, 0.25182274999831583]), ('F1', [0.7675675675675676, 0.7675675675675676, 0.9594594594594594, 0.9589041095890412, 0.9655172413793103, 0.9655172413793103, 0.9722222222222222, 0.9722222222222222, 0.9722222222222222, 0.9722222222222222])])}
```

If you have `pandas` available and would prefer to work with the metrics in that format, you can do the following.

```python
import pandas as pd
metrics_df = pd.DataFrame.from_records(lgb_model.evals_result_["valid"])
metrics_df
```

```text
         F1       auc  binary_error  binary_logloss
0  0.767568  0.956436      0.377193        0.582173
1  0.767568  0.972977      0.377193        0.514326
2  0.959459  0.971667      0.052632        0.462733
3  0.958904  0.972650      0.052632        0.415805
4  0.965517  0.972650      0.043860        0.377370
5  0.965517  0.967082      0.043860        0.347271
6  0.972222  0.966754      0.035088        0.319601
7  0.972222  0.967245      0.035088        0.294086
8  0.972222  0.994432      0.035088        0.268995
9  0.972222  0.994104      0.035088        0.251823
```

For docs on all of this, see:

* https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMClassifier.html
* https://lightgbm.readthedocs.io/en/latest/Parameters.html

For more discussion you might find relevant, see #3029.
